src/ray/core_worker/transport/direct_actor_transport.h

```diff
@@ -265,20 +265,53 @@ class FiberEvent {
   bool ready_ = false;
 };
 
+class FiberSemaphore {
+  public:
+    FiberSemaphore(int num): num_(num) {
+      RAY_LOG(INFO) << "Initialized fiber semphaore with concurrency " << num;
+    }
+    
+    // Enter the semaphore. Wait fo the value to be > 0 and decrement the value.
+    void Down() {
+      RAY_LOG(INFO) << "Semphoare down requested";
+      std::unique_lock<boost::fibers::mutex> lock(mutex_);
+      cond_.wait(lock, [this]() { return num_ > 0; });
+      num_ -= 1;
+      RAY_LOG(INFO) << "Semphoare down complete";
+    }
+
+    // Exit the semaphore. Increment the value and notify other waiter.
+    void Up() {
+      RAY_LOG(INFO) << "Semphoare up";
+      {
+        std::unique_lock<boost::fibers::mutex> lock(mutex_);
+        num_ += 1;
+      }
+      cond_.notify_one();
+    }
```

Comment:
One downside of this is that you don't provide any ordering on how tasks are admitted. This could be a problem for certain workloads, so consider adding a TODO.